lib/files.js

```diff
@@ -35,22 +39,90 @@ function createImportHook() {
  */
 function createSyncImportHook() {
   let imports = {};
-  return function (fpath, kind, cb) {
-    fpath = path.resolve(fpath);
+  return function ({path: fpath, importerPath}, cb) {
+    fpath = path.resolve(path.dirname(importerPath), fpath);
     if (imports[fpath]) {
       cb();
     } else {
       imports[fpath] = true;
-      cb(null, fs.readFileSync(fpath, {encoding: 'utf8'}));
+      cb(null, {
+        contents: fs.readFileSync(fpath, {encoding: 'utf8'}),
+        path: fpath
+      });
     }
   };
 }
 
+/**
+ * Convenience function to parse multiple inputs into protocols and schemas.
+ *
+ * It should cover most basic use-cases but has a few limitations:
+ *
+ * + It doesn't allow passing options to the parsing step.
+ * + The protocol/type inference logic can be deceived.
+ *
+ * The parsing logic is as follows:
+ *
+ * + If `str` contains `path.sep` (on windows `\`, otherwise `/`) and is a path
+ *   to an existing file, it will first be read as JSON, then as an IDL
+ *   specification if JSON parsing failed. If either succeeds, the result is
+ *   returned, otherwise the next steps are run using the file's content
+ *   instead of the input path.
+ * + If `str` is a valid JSON string, it is parsed then returned.
+ * + If `str` is a valid IDL protocol specification, it is parsed and returned
+ *   if no imports are present (and an error is thrown if there are any
+ *   imports).
+ * + If `str` is a valid IDL type specification, it is parsed and returned.
+ * + If neither of the above cases apply, `str` is returned.
+ */
+function readSchemaFromPathOrString(str) {
+  let schema;
+  if (
+    typeof str == 'string' &&
+    str.indexOf('/') !== -1 &&
+    fs.existsSync(str)
+  ) {
+    // Try interpreting `str` as path to a file contain a JSON schema or an IDL
+    // protocol. Note that we add the second check to skip primitive references
+    // (e.g. `"int"`, the most common use-case for `avro.parse`).
+    let contents = fs.readFileSync(str, {encoding: 'utf8'});
```

Comment:
This is the only part that requires `fs` - could we trim this function only include this logic? We could call it something like `tryReadFileSync`. The change would avoid the `./spec` import here and logic duplication between the two files.